### PR TITLE
Change priority subject determination for AB#16772.

### PR DESCRIPTION
**Pull request by BrianMaki**
# Fixes [AB#16772](https://dev.azure.com/qslvic/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/16772)

## Description

Change priority for application subject determination:

- Check if Hdid exists in route values
- Check if Hdid exists in query parameters

Created common HttpContextHelper class with common code to access Hdid Route and Parameter Values:

- Moved GetResourceHdid from BaseFhirAuthorizationHandler to HttpContextHelper
- Created unit test for HttpContextHelper

Priority:

1. Hdid in route
2. Hdid in query parameter
3. Hdid in claim identity
4. Subject PHN
5. Subject query

Also updated unit tests.

Screenshot 2024-06-24 at 12 56 37 PM

## Testing

- [x] Unit Tests Updated
- [ ] Functional Tests Updated
- [ ] Not Required

## UI Changes

None

## Items to Review:

-   [General PR Guidelines](https://github.com/bcgov/healthgateway/wiki/PRguidance)

### CODE REVIEW

**Review by MikeLyttle — CHANGES_REQUESTED** [review]
The `subjectQuery` variable is coming from a request parameter with the name "queryString" - we usually are passing the HDID in via a parameter called "hdid".  Though sometimes it comes as part of the route instead of as a query string.

File: Apps/Common/src/Auditing/AuditLogger.cs [hunk 1]
```diff
@@ -57,7 +57,7 @@ public void PopulateWithHttpContext(HttpContext context, AuditEvent auditEvent)
             auditEvent.ApplicationType = GetApplicationType();
             auditEvent.TransactionResultCode = GetTransactionResultType(context.Response.StatusCode);
 
-            auditEvent.ApplicationSubject = hdid ?? subjectPhn ?? subjectQuery;
+            auditEvent.ApplicationSubject = subjectQuery ?? hdid ?? subjectPhn;
```
Comment (sslaws) [inline Apps/Common/src/Auditing/AuditLogger.cs:60]: As Mike indicated logic needs to be made smarter for the APIs (not for Admin) and should check route values and query parms for HDID.

**Review by sslaws — CHANGES_REQUESTED** [review]

File: Apps/Common/src/Auditing/AuditLogger.cs (hunk 1 above)
Comment (BrianMaki) [reply Apps/Common/src/Auditing/AuditLogger.cs]: Route values and query parameters are now checked for Hdid.

File: Apps/Common/src/Auditing/AuditLogger.cs [hunk 2]
```diff
@@ -57,10 +57,17 @@ public void PopulateWithHttpContext(HttpContext context, AuditEvent auditEvent)
             auditEvent.ApplicationType = GetApplicationType();
             auditEvent.TransactionResultCode = GetTransactionResultType(context.Response.StatusCode);
 
-            auditEvent.ApplicationSubject = hdid ?? subjectPhn ?? subjectQuery;
+            // Check if Hdid is in route values
+            RouteValueDictionary routeValues = context.Request.RouteValues;
+            string? routeHdid = routeValues["Hdid"] as string;
+
+            // Check if Hdid is in query parameters
+            context.Request.Query.TryGetValue("Hdid", out StringValues queryHdidParameter);
```
Comment (MikeLyttle) [inline Apps/Common/src/Auditing/AuditLogger.cs:65]: I had to double-check to verify these retrievals were both case-insensitive, but they are.

File: Apps/Common/test/unit/Auditing/DbAuditLoggerTests.cs [hunk 3]
```diff
@@ -240,5 +234,51 @@ public async Task ShouldWriteAuditEventHandleException()
                     It.IsAny<Func<It.IsAnyType, Exception?, string>>()),
                 Times.Once);
         }
+
+        private static GetPopulateWithHttpContextMock SetupGetPopulateWithHttpContextMock(bool useRouteValues, bool useQueryParamValues)
```
Comment (MikeLyttle) [inline Apps/Common/test/unit/Auditing/DbAuditLoggerTests.cs:238]: Nit: I think the type should be named `PopulateWithHttpContextMock` and the method name should be either `SetupPopulateWithHttpContextMock` or `GetPopulateWithHttpContextMock`.

File: Apps/Common/src/Auditing/AuditLogger.cs (hunk 2 above)
Comment (BrianMaki) [reply Apps/Common/src/Auditing/AuditLogger.cs:65]: Route values are case sensitive but query parameters are not.

**Review by MikeLyttle — APPROVED** [review]

**Review by MikeLyttle — APPROVED** [review]

**Review by sslaws — APPROVED** [review]

**Pull request merged**